src/FieldMapB.cpp

```diff
@@ -0,0 +1,377 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Wouter Deconinck, Dhevan Gangadharan, Aranya Giri
+
+#include <DD4hep/DetFactoryHelper.h>
+#include <DD4hep/FieldTypes.h>
+#include <DD4hep/Printout.h>
+#include <XML/Utilities.h>
+
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+namespace fs = std::filesystem;
+
+#include "FileLoaderHelper.h"
+
+using namespace dd4hep;
+
+//  Two coordinate options:
+//  coord_type="BrBz"
+//    expected fields contained within <dimensions>:
+//          <R step="" min="" max=""/>
+//          <Z step="" min="" max=""/>
+//  coord_type="BxByBz"
+//    expected fields contained within <dimensions>:
+//          <X step="" min="" max=""/>
+//          <Y step="" min="" max=""/>
+//          <Z step="" min="" max=""/>
+
+// In DD4hep 1.26, the name for the `field_type` enum changed (from `type`).
+#if DD4HEP_VERSION_GE(1, 26)
+#define DD4HEP_FIELD_TYPE field_type
+#else
+#define DD4HEP_FIELD_TYPE type
+#endif
+
+// This allows us to specify the name of the variable by hand, when patching
+// the previous versions, by setting `DD4HEP_FIELD_TYPE_OVERRIDE`.
+// TODO remove after DD4hep 1.26 release
+#define XSTR(x) STR(x)
+#define STR(x) #x
+#ifdef DD4HEP_FIELD_TYPE_OVERRIDE
+#undef DD4HEP_FIELD_TYPE
+#define DD4HEP_FIELD_TYPE DD4HEP_FIELD_TYPE_OVERRIDE
+#pragma message("DD4HEP_FIELD_TYPE overridden as " XSTR(DD4HEP_FIELD_TYPE))
+#endif
+#pragma message("DD4HEP_FIELD_TYPE is " XSTR(DD4HEP_FIELD_TYPE))
+
+// implementation of the field map
+class FieldMapB : public dd4hep::CartesianField::Object {
+public:
+  FieldMapB(const std::string& field_type_str = "magnetic", const std::string& coord_type_str = "BrBz");
+  void Configure(std::vector<xml_comp_t> dimensions);
+  void LoadMap(const std::string& map_file, double scale);
+  bool GetIndices(double R, double Z, int *idxR, int *idxZ, double *deltaR, double *deltaZ);
+  bool GetIndices(double X, double Y, double Z, int *idxX, int *idxY, int *idxZ, double *deltaX, double *deltaY, double *deltaZ);
+  void SetTransform(const Transform3D& tr)
+  {
+    trans     = tr;
+    trans_inv = tr.Inverse();
+  }
+
+  virtual void fieldComponents(const double* pos, double* field);
+
+private:
+  bool                                                          case_BrBz; // is it BrBz or not (BxByBz)
```

Comment:
Can we use `enum` and switch case (the "case" hints at that too)?

Reply:
Good idea.  done.